File: WordBooks/app/src/main/java/com/callor/word/model/WordViewModel.java
package com.callor.word.model;
/*
WordVO : 데이터를 담고, Table 을 생성하는 용도의 클래스
WordViewModel : ViewModel 클래스로 데이터를
Observe 방식으로 핸들링 하기 위한 도구
 */

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.callor.word.persistance.WordRepository;

import java.util.List;

public class WordViewModel extends ViewModel {

    private WordRepository wordRepository;

    public WordViewModel(Application application) {
        wordRepository = new WordRepository(application);
    }

    public LiveData<List<WordVO>> selectAll() {
        return wordRepository.wordList();
    }

    public void insert(WordVO wordVO) {
        wordRepository.insert(wordVO);
    }
}

File: WordBook/app/src/main/java/com/callor/word/model/WordViewModel.kt
package com.callor.word.model

import androidx.lifecycle.*
import com.callor.word.persistance.WordRepository
import kotlinx.coroutines.launch

class WordViewModel(private val repository: WordRepository) : ViewModel(){

    // Flow 데이터를 LiveData 로 변경
    // 한번 만들어둔 view 화면에서 데이터가 바뀌면 자동으로
    // 바뀐 데이터를 보여주게 하는 코드
    val allWords : LiveData<List<Word>> = repository.allWords.asLiveData()
    fun insert(word:Word) = viewModelScope.launch {
        repository.insert(word)
    }

}

class WordViewModelFactory(private val repository: WordRepository) 
    : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if(modelClass.isAssignableFrom(WordViewModel::class.java)) {
            return WordViewModel(repository) as T
        }
        throw IllegalAccessException("View Model 클래스 없음")
    }
}


